amqp/sasl.py

```diff
@@ -36,6 +36,8 @@ def __init__(self, username, password):
         self.username, self.password = username, password
 
     def start(self, connection):
+        if self.username is None or self.password is None:
+            return NotImplemented
```

Comment:
Why are we returning NotImplemented? It doesn't seem like it's the right thing to do.
See https://stackoverflow.com/a/44575926/85140

Reply:
I had reviewed #110, more specifically the comment of 5f63857f6af350309652bd2ee19a3b6531af33eb, and thought I should adjust to that. However I do agree it does not seem right. What alternatives would you prefer?

One option would be, if credentials are missing, not to add them to `self.authentication` in the first place. In that case I would prefer moving the argument availability checks to the `SASL` classes entirely so that the `__init__()` method in `Connection` does not get too complex.

The second option would simply be to return something else than `NotImplemented`, e.g. `None` or a custom exception during `start()`.

Reply:
If we were to change that, I'd do so in a different PR. I just didn't know that's the way it currently works.